Enable Confidential Containers on Baremetal

beraldoleal opened the pull request:
This adds initial support for Confidential on baremetal. Until now, Confidential was only supported by peerpods. With this change, we are introducing the Confidential feature to baremetal as well.
    
Our consensus so far, is to use the same config option "confidential" to enable Confidential features in both cases.
    
We are also reusing the same logic for KataConfigPoolSelector and checkNodeEligibility, just as we have been doing so far.

One aspect we are evaluating is enabling the MachineConfigs on all nodes alongside CRIO, regardless of whether the machine has support. The RuntimeClass would be responsible for filtering machines according to their capabilities.

RPM work is still needed.

beraldoleal committed:
controllers: add additional parameters

This method does not support additional filters and custom handler. This
is prep-work for future commits. The upcoming kata-cc runtime class will
be using different TEE-specific handlers.

Signed-off-by: Beraldo Leal <bleal@redhat.com>

beraldoleal commented:
@gkurz , while still a working in progress, feel free to validate this approach with our discussions. Also keep in mind MachineConfig files are just boilerplate at this moment as example.

beraldoleal commented:
@bpradipt @gkurz do you guys see any issues with this direction?

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -2405,26 +2409,15 @@ func (r *KataConfigOpenShiftReconciler) createAuthJsonSecret() error {
-		r.Log.Info("Error in creating kata remote configuration.toml MachineConfig", "err", err)
-		return err
+// Create MachineConfigs from a list of YAML files
+func (r *KataConfigOpenShiftReconciler) createMachineConfigsFromFiles(yamlFiles []string) error {
```
I don't find this function very useful actually... in the end it just calls `createMcFromFile` for a list of files that the caller must provide. What's the goal ?

If it is to reduce the number of LoC, I'd rather consolidate the `filepath.join` in `createMcFromFile` and add an  extra location path argument  and have a `createMcFromFile(someMachineConfigPathLocation, someMachineConfigYaml)` for all MC that need to be created instead of the array looping.

gkurz replied:
I'm actually not a bit fan of having so many MachineConfigs to deploy the CRI-O config files. They have a fixed content and could be easily shipped with the RPM instead of all the boiler plate to do that from the controller.

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -2426,6 +2449,66 @@ func (r *KataConfigOpenShiftReconciler) createMachineConfigsFromFiles(yamlFiles
+			},
+		}
+
+		err := r.Client.Delete(context.TODO(), rc)
```
This isn't needed because `createRuntimeClass` does :

```
	// Set Kataconfig r.kataConfig as the owner and controller
	if err := controllerutil.SetControllerReference(r.kataConfig, rc, r.Scheme); err != nil {
		return err
	}
```

This causes the RuntimeClass object to be automatically deleted when KataConfig goes away.

beraldoleal replied:
For public visibility, we decided to remove the runtimeclass when confidential is disabled. This ensures users cannot deploy confidential pods. The same applies to kata-remote and any other runtimeclass except kata.

beraldoleal committed:
controllers: create kata-cc-* runtime classes

This adds initial support for Confidential on baremetal. Until now,
Confidential was only supported by peerpods. With this change, we are
introducing the Confidential feature to baremetal as well.

Our consensus is to use the same config option "confidential" to enable
Confidential features in both cases.

We are also reusing the same logic for KataConfigPoolSelector and
checkNodeEligibility, just as we have been doing so far.

Currently, intel and amd labels uses the same string as defined at
scripts/install-helpers/baremetal-coco/nfd/intel-rules.yaml#L26 and
scripts/install-helpers/baremetal-coco/nfd/amd-rules.yaml#L10. Ideally
we should be using a NFD standard string for that.

All the CRI-O and kata-containers configurations will be installed by
default on all nodes , regardless of whether the machine has support.
This will be RPM responsability. The RuntimeClass would be responsible
for filtering machines according to their capabilities.

Related: KATA-3967.

Signed-off-by: Beraldo Leal <bleal@redhat.com>

bpradipt commented on controllers/confidential_handler.go:
```diff
@@ -57,19 +71,49 @@ func (r *KataConfigOpenShiftReconciler) handleFeatureConfidential(state FeatureG
+		}
+
+		// Create kata-cc-amd runtime class for AMD SNP
+		err = r.createRuntimeClass(kataCCAMDRuntimeClassName, kataCCAMDRuntimeClassCpuOverhead, kataCCAMDRuntimeClassMemOverhead, amdSNPNodeLabel)
```
Do we really want to create arch specific runtimeClass or just use `kata-cc` as the runtimeClass with handlers as `kata-cc-{arch}` ?

beraldoleal replied:
I’m ok with doing a single RC (kata-cc), but iiuc we can’t have multiple handlers since a RuntimeClass only takes one. So if we stick with one RC, the RPM would need to handle it by dropping all the configs and setting up the right configuration.toml (or symlink) depending on the platform.

Ex: on a TDX node it points to configuration-qemu-tdx.toml, on an SNP node it points to configuration-qemu-snp.toml. That way kata-cc always just uses the kata handler, and the node decides which TEE to run.

Users could still use node affinity if they want to pin a pod to Intel vs AMD.

bpradipt replied:
> Ex: on a TDX node it points to configuration-qemu-tdx.toml, on an SNP node it points to configuration-qemu-snp.toml. That way kata-cc always just uses the kata handler, and the node decides which TEE to run.

Right. This was in my mind. The premise was that this is easier from an end-user standpoint deploying confidential workload in the cluster. Wdyt ?

beraldoleal replied:
I'm ok with that approach. 

However, @gkurz mentioned something about having multiple runtimeclasses but also having a single one to be the "smart" one. Its still not clear to me, if we really need those, if the users can choose the nodes by using the node affinity labels.

So I will just wait a bit for his comments, to see if he has something against this.

gkurz replied:
I'm now thinking that the operator should focus on a single "smart" runtime class `kata-cc` for a better end user experience. If the admin wants to add TEE-specific runtime classes, they can still do it manually since the RPM provides the CRI-O handlers. We can document how to do that.

beraldoleal replied:
Thanks @gkurz . As we agreed today during the call, I will change this PR to use a single runtime class, However, since this is for the TP in November, we are assuming for now this will handle only single TEE per cluster, not multiple TEE platforms on the same cluster.

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -880,7 +880,7 @@ func (r *KataConfigOpenShiftReconciler) createScc() error {
 }
 
-func (r *KataConfigOpenShiftReconciler) createRuntimeClass(runtimeClassName string, cpuOverhead string, memoryOverhead string) error {
+func (r *KataConfigOpenShiftReconciler) createRuntimeClass(runtimeClassName string, cpuOverhead string, memoryOverhead string, additionalNodeLabel string) error {
```
Since the additional label will come from NFD, I'm wondering if this is the right way to go. Shouldn't we rather adapt our existing CheckNodeEligibility logic ?

beraldoleal replied:
IIUC, if TEE label logic moves into CheckNodeEligibility, only TEE nodes are eligible for the Kata install.

With the current approach, the pool stays broad and the RuntimeClass narrows via TEE label, so kata, kata-remote and kata-cc can coexist.

Does that make sense? Or you have something else in mind?

gkurz replied:
I see CheckNodeEligibility as way to indicate that our runtime classes should take node properties into account, otherwise we assume that the admin is responsible to use kata on proper HW.

Each runtime class would have its specific selector :

- kata-remote : can run everywhere where kata was deployed (kata-oc)
- kata : can run on kata-oc hosts where KVM is available
- kata-cc: can run on kata-oc hosts where KVM  and TEE are available

We might have to come up with some NFD rules to make this fully functional.

beraldoleal replied:
Yes, I think CheckNodeEligibility was added for the simpler (non‑confidential) case, so it probably needs a revisit now that TEE is in play.

My concern is that ... IIUC, currently it gates the kata installation pool (who gets node-role.kubernetes.io/kata-oc) AND, because our RuntimeClasses select kata-oc, it also constrains where RCs can run indirectly.

If we fold TEE into CheckNodeEligibility, won't we be shrinking the Kata installation footprint to TEE nodes only? Don't we need to think also in terms of kata installation as well?

beraldoleal replied:
As discussed on slack, we can ignore this for now. closing the thread.

beraldoleal commented:
@bpradipt @wainersm @gkurz ready for a second round. let me know if you have any comments.

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -1481,6 +1509,33 @@ func (r *KataConfigOpenShiftReconciler) processKataConfigInstallRequest() (ctrl.
 }
 
+// Delete runtime class
+func (r *KataConfigOpenShiftReconciler) deletRuntimeClass(runtimeClassName string) error {
```
Left over ?

gkurz replied:
The function is still around...

gkurz commented on controllers/confidential_handler.go:
```diff
@@ -57,22 +77,98 @@ func (r *KataConfigOpenShiftReconciler) handleFeatureConfidential(state FeatureG
-			return err
+		r.Log.Info("Deleting kata-cc runtime classes for confidential containers")
+
+		// Delete kata-cc runtime class
```
I'm still a bit concerned that we can delete the runtime class even though it might still be used. Currently running pods will be unaffected but it will prevent new pods from being created using that runtime class and can cause instability during rollouts or scaling events in deployments and daemonsets... This badly breaks the end user experience IMO.

beraldoleal replied:
I see the usability issue you mentioned. A few alternatives we could consider:

1. Add a pod validation webhook to reject pods using kata-cc when confidential is disabled. Not sure about the operational implications of adding another webhook though, specially for every new pod. I feel it adds some overhead we don't need.

2. Keep the runtime class but update the handler to "kata". I don't like this either, it silently masks the issue and users might think they're getting confidential computing when they're not.

3. Update the existing KataConfig validation webhook to block disabling confidential if there are pods running with kata-cc or kata-remote. This would force users to migrate workloads first.

Wdyt?

beraldoleal replied:
As we discussed today, I implemented the finalizer approach. Vide last commits and let me know if you spot any issues.

beraldoleal committed:
controllers: add RuntimeClass finalizer management

Implement controller to prevent deletion of RuntimeClasses with active
pods.

This is valid for all runtime classes created by this controller with
createRuntimeClass().

This ensures that RuntimeClasses cannot be deleted while pods are still
dependent on them, preventing potential cluster instability.

Signed-off-by: Beraldo Leal <bleal@redhat.com>

beraldoleal committed:
controllers: add migration for finalizers

Add ensureRuntimeClassFinalizers() to migration logic to retroactively
add finalizers to existing runtime classes created before the finalizer
implementation.

Signed-off-by: Beraldo Leal <bleal@redhat.com>

beraldoleal committed:
rabc: make manifests and make bundle

Since we introduced a controller, I just ran make manifests and make
bundle.

Signed-off-by: Beraldo Leal <bleal@redhat.com>

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -783,7 +783,7 @@ func (r *KataConfigOpenShiftReconciler) createRuntimeClass(runtimeClassName stri
 				Name: runtimeClassName,
 			},
-			Handler: runtimeClassName,
+			Handler: handler,
```
Maybe worth clarifying in the commit message the need to do this, e.g. "the upcoming kata-cc runtime class will be using different TEE-specific handlers"

gkurz commented on controllers/confidential_handler.go:
```diff
@@ -1,7 +1,27 @@
+
+	// TEE node labels
+	intelTDXNodeLabel = "intel.feature.node.kubernetes.io/tdx"
+	amdSNPNodeLabel   = "amd.feature.node.kubernetes.io/snp"
```
Where do these labels come from ?

According to https://kubernetes-sigs.github.io/node-feature-discovery/v0.12/usage/features.html#cpu , if the NFD operator is deployed, we should get the following :

```
feature.node.kubernetes.io/cpu-security.tdx.enabled
feature.node.kubernetes.io/cpu-security.sev.snp.enabled	
```

beraldoleal replied:
Yeah, I'm not sure if we have any good reason for not using them, and probably we should use the standard NFD labels, at some point.

However, right now, I'm just reusing the vendor-prefixed labels (`intel.feature.node.kubernetes.io/tdx`, and `amd.feature.node.kubernetes.io/snp`) that are already in our codebase (`scripts/install-helpers/baremetal-coco/nfd/`).

I would like to keep this PR using the same labels as we use already use today and let this discussion for another PR, if possible.

gkurz replied:
That's fine but please add a comment that the strings should exactly match
https://github.com/openshift/sandboxed-containers-operator/blob/devel/scripts/install-helpers/baremetal-coco/nfd/intel-rules.yaml#L26 and https://github.com/openshift/sandboxed-containers-operator/blob/devel/scripts/install-helpers/baremetal-coco/nfd/amd-rules.yaml#L10 for now.

mythi replied:
These `NodeFeatureRules` are important since they also register the necessary node resources that the runtimeclasses here need to add as the overhead values.

beraldoleal commented:
@bpradipt @gkurz , please let me know if we are ok to merge this one.

bpradipt commented on controllers/confidential_handler.go:
```diff
@@ -1,7 +1,27 @@
+)
+
+const (
+	// kata-cc runtime class for CoCo
```
nit
`s/CoCo/CoCo BM`

gkurz commented:
In the log of commit 2 :
> All the MachineConfigs will be installed by default on all nodes

I guess you meant `All the CRI-O and kata-containers configurations will be installed by default on all nodes` since we're moving away from MachineConfigs.

gkurz commented on controllers/confidential_handler.go:
```diff
@@ -57,22 +77,98 @@ func (r *KataConfigOpenShiftReconciler) handleFeatureConfidential(state FeatureG
+		}
+	}
+
+	if hasIntelTDX && hasAmdSNP && !r.kataConfig.Spec.EnablePeerPods {
```
Not sure to understand the goal of `!r.kataConfig.Spec.EnablePeerPods` here...

beraldoleal replied:
You are right, makes no sense. I removed.

beraldoleal commented:
@gkurz @bpradipt sent the changes, please take a look when possible.

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -782,6 +782,7 @@ func (r *KataConfigOpenShiftReconciler) createRuntimeClass(runtimeClassName stri
 			},
 			ObjectMeta: metav1.ObjectMeta{
 				Name: runtimeClassName,
```
```suggestion
				Name:       runtimeClassName,
```

to silent `make fmt` .

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -782,6 +782,7 @@ func (r *KataConfigOpenShiftReconciler) createRuntimeClass(runtimeClassName stri
 			},
 			ObjectMeta: metav1.ObjectMeta{
 				Name: runtimeClassName,
+				Finalizers: []string{"kataconfig.openshift.io/runtime-class-cleanup"},
```
```suggestion
				Finalizers: []string{runtimeClassFinalizerName},
````

gkurz commented on controllers/migrate.go:
```diff
@@ -81,3 +84,32 @@ func (r *KataConfigOpenShiftReconciler) migratePeerPodsLimit() error {
+			return err
+		}
+
+		if !controllerutil.ContainsFinalizer(rc, "kataconfig.openshift.io/runtime-class-cleanup") {
```
```suggestion
		if !controllerutil.ContainsFinalizer(rc, runtimeClassFinalizerName) {
```

gkurz commented on controllers/migrate.go:
```diff
@@ -81,3 +84,32 @@ func (r *KataConfigOpenShiftReconciler) migratePeerPodsLimit() error {
+
+		if !controllerutil.ContainsFinalizer(rc, "kataconfig.openshift.io/runtime-class-cleanup") {
+			r.Log.Info("Adding finalizer to existing runtime class", "runtimeClass", name)
+			controllerutil.AddFinalizer(rc, "kataconfig.openshift.io/runtime-class-cleanup")
```
```suggestion
			controllerutil.AddFinalizer(rc, runtimeClassFinalizerName)
```

gkurz commented on config/rbac/role.yaml:
This change should be propagated to the bundle, i.e. `make bundle`.

gkurz commented on controllers/confidential_handler.go:
```diff
@@ -57,22 +77,98 @@ func (r *KataConfigOpenShiftReconciler) handleFeatureConfidential(state FeatureG
 	}
 
 	// peer pods config
+	if r.kataConfig.Spec.EnablePeerPods {
```
All the lines above are also specific to peer pods AFAICT. It seems this `if` should move to line 40 for now.

gkurz replied:
No longer getting the errors from the image generator. Fixed.

gkurz commented:
I deployed this PR in my cluster and did the following :
- start a regular kata POD
- delete the kata runtime class
- observe that `oc delete` remains stuck
-  see the following trace in the logs of the controller :
```
2025-10-07T15:22:43Z	INFO	controllers.RuntimeClass	RuntimeClass is being deleted, checking for active pods	{"runtimeclass": "kata"}
2025-10-07T15:22:43Z	INFO	controllers.RuntimeClass	Found active pod using runtime class	{"pod": "example-fedora-kata", "namespace": "supergroug", "runtimeClass": "kata", "phase": "Running"}
2025-10-07T15:22:43Z	INFO	controllers.RuntimeClass	Runtime class still has active pods, waiting for them to be deleted manually or terminate naturally	{"runtimeclass": "kata"}
```
- delete the kata POD
- observe that `oc delete` completes
-  see the following trace in the logs of the controller :
```
2025-10-07T15:23:13Z	INFO	controllers.RuntimeClass	No active pods found, removing finalizer to allow deletion	{"runtimeclass": "kata"}
2025-10-07T15:23:13Z	INFO	controllers.RuntimeClass	Successfully removed finalizer, RuntimeClass will be deleted	{"runtimeclass": "kata"}
```

The finalizer-based logic to protect active runtime classes works as expected.

I could also verify the migration logic on a previous deployment :
```
2025-10-07T14:15:03Z	INFO	controllers.KataConfig	Adding finalizer to existing runtime class	{"runtimeClass": "kata"}
```

This is promising ! :smile:

gkurz reviewed (approved):
/lgtm

Thanks @beraldoleal !

mythi commented on controllers/confidential_handler.go:
```diff
@@ -18,61 +38,136 @@ import (
+		}
+
+		// Create kata-cc runtime class restricted to the detected TEE subset
+		err = r.createRuntimeClass(kataCCRuntimeClassName, kataCCRuntimeClassCpuOverhead, kataCCRuntimeClassMemOverhead, handler, nodeLabel)
```
in addition to CPU and Mem overhead, we consume "key ID" overhead resources, one for each CVM.

beraldoleal replied:
I'm not sure if we should handle this inside the RuntimeClass or would it need to be a different mechanism. IIUC, it seems like a different type of resource.

Since this isn't handled today, I think it should be addressed in a separate PR.
Could you expand a bit on this key ID overhead?

gkurz replied:
@mythi can you share a link to the definition of this "key ID" resource ?

mythi replied:
Pod overhead accepts any kubernetes resource, including "extended resources". We have NFD configured to register cvm keys resources as node extended resources and they are used in pod overhead https://github.com/openshift/sandboxed-containers-operator/blob/c8c9e5eaf8c24d30bc5915cf83548f1e9b846b0c/scripts/install-helpers/baremetal-coco/install.sh#L532 This is the bookkeeping to ensure we don't run more TEE pods than what the node is capable of

gkurz replied:
I suggest we merge this PR and we add support for this in a subsequent one. Is it ok for you @mythi ?

mythi replied:
It's ok

mythi replied:
a follow-up PR would also need to cleanup `install.sh` for the parts added here, right?

wainersm commented on controllers/confidential_handler.go:
```diff
@@ -18,61 +38,136 @@ import (
 // When the feature is disabled, handleFeatureConfidential resets the config maps to non-confidential values.
 func (r *KataConfigOpenShiftReconciler) handleFeatureConfidential(state FeatureGateState) error {
```
Hi @beraldoleal ! `handleFeatureConfidential()` doc string should be updated.

wainersm commented on controllers/confidential_handler.go:
```diff
@@ -18,61 +38,136 @@ import (
+				return err
+			}
+		}
+	}
```
It will handle CC on peer-pods `if r.kataConfig.Spec.EnablePeerPods` and if it reached to this point it should `return` otherwise it will continue the execution and try to run the code block following. In other words, it should not try to setup the `kata-cc` runtimeclass if peer pods. Did I get correct?

While in here, maybe I would split the handler code for baremetal and peer-pods in two functions (say `handle_baremetal()` and `handle_peerpods()`) and make the  `handleFeatureConfidential` only deciding which handler to call.

beraldoleal replied:
> It will handle CC on peer-pods if r.kataConfig.Spec.EnablePeerPods and if it reached to this point it should return otherwise it will continue the execution and try to run the code block following. In other words, it should not try to setup the kata-cc runtimeclass if peer pods. Did I get correct?

From what we discussed before... peerpods and baremetal can coexist. Users can choose which to use via the RuntimeClass in their pod spec. We so need to handle both configs here.

> While in here, maybe I would split the handler code for baremetal and peer-pods in two functions (say handle_baremetal() and handle_peerpods()) and make the handleFeatureConfidential only deciding which handler to call.

That is a good suggestion. I will push the split.

wainersm replied:
> > It will handle CC on peer-pods if r.kataConfig.Spec.EnablePeerPods and if it reached to this point it should return otherwise it will continue the execution and try to run the code block following. In other words, it should not try to setup the kata-cc runtimeclass if peer pods. Did I get correct?
> 
> From what we discussed before... peerpods and baremetal can coexist. Users can choose which to use via the RuntimeClass in their pod spec. We so need to handle both configs here.
> 

Ok, I overlooked that. Even though, if you let the execution continue and it call `handler, nodeLabel, err := r.computeTEEHandlerAndLabel()`  it will fail if you have only peer-pods CoCo. 

> > While in here, maybe I would split the handler code for baremetal and peer-pods in two functions (say handle_baremetal() and handle_peerpods()) and make the handleFeatureConfidential only deciding which handler to call.
> 
> That is a good suggestion. I will push the split.

beraldoleal replied:
> Ok, I overlooked that. Even though, if you let the execution continue and it call handler, nodeLabel, err := r.computeTEEHandlerAndLabel() it will fail if you have only peer-pods CoCo.

You are right.

What I just pushed is the following logic when confidential is enabled and we fail to find TEE hardware:

* If peer pods is enabled, logs a warning and skips baremetal CoCo (because peer pods CVM will handle confidential workloads)
* If peer pods is disabled, returns an error (because admin explicitly requested confidential but there's no hardware to support it)

Does that makes sense for now? let me know what you think

wainersm replied:
Hi @beraldoleal , yes, it makes sense to me. Thanks for addressing that too!

wainersm commented on controllers/confidential_handler.go:
```diff
@@ -18,61 +38,136 @@ import (
+		// Create kata-cc runtime class restricted to the detected TEE subset
+		err = r.createRuntimeClass(kataCCRuntimeClassName, kataCCRuntimeClassCpuOverhead, kataCCRuntimeClassMemOverhead, handler, nodeLabel)
+		if err != nil {
+			r.Log.Info("Error creating kata-cc runtime class", "err", err)
```
Use the constant `kataCCRuntimeClassName` in the log messages.

I don't know if it should be "runtime class" or "runtimeclass" or "runtimeClass". We just need to keep it consistent because people grep those logs for debugging problems.

wainersm commented on controllers/confidential_handler.go:
```diff
@@ -18,61 +38,136 @@ import (
+	// Handle kata-cc runtime classes for baremetal environments
 	if state == Enabled {
-		peerpodsCMData = map[string]string{"DISABLECVM": "false"}
+		r.Log.Info("Creating kata-cc runtime classes for confidential containers")
```
Use `kataCCRuntimeClassName` (see comment below)

wainersm commented on controllers/confidential_handler.go:
```diff
@@ -18,61 +38,136 @@ import (
+
+	nodes := &corev1.NodeList{}
+	listOpts := []client.ListOption{
+		client.MatchingLabelsSelector{Selector: selector},
```
`getKataConfigNodeSelectorAsSelector()` might return an empty map as seen in https://github.com/openshift/sandboxed-containers-operator/blob/devel/controllers/openshift_controller.go#L844 . Is it okay having `MatchingLabelsSelector` with a empty map?

beraldoleal replied:
iiuc, an empty selector is valid here, it will return everything. And it just means the admin wants to install on all nodes cluster-wide.

We use the same pattern in `getNodesWithLabels()`.

wainersm commented on controllers/confidential_handler.go:
```diff
@@ -18,61 +38,136 @@ import (
+		return "", "", fmt.Errorf("%s", msg)
+	}
+
+	if hasIntelTDX && !hasAmdSNP {
```
Checking `!hasAmdSNP` is not needed as both cannot be set at this point.

beraldoleal replied:
Good catch. Fixing..

wainersm commented on controllers/confidential_handler.go:
```diff
@@ -18,61 +38,136 @@ import (
+	if hasIntelTDX && !hasAmdSNP {
+		return kataCCIntelHandler, intelTDXNodeLabel, nil
+	}
+	if hasAmdSNP && !hasIntelTDX {
```
Same here.

wainersm commented on controllers/confidential_handler.go:
```diff
@@ -18,61 +38,136 @@ import (
+func (r *KataConfigOpenShiftReconciler) computeTEEHandlerAndLabel() (string, string, error) {
+	selector, err := r.getKataConfigNodeSelectorAsSelector()
+	if err != nil {
+		msg := "Couldn't build KataConfig node selector for TEE detection"
```
In this function I've seen this pattern: log a message & return the error. However, the caller function will also log a message (that contain the message already logged) and return the error. It will produce duplicated log messages.

beraldoleal replied:
Good catch, thanks. Fixing.

wainersm commented on controllers/runtimeclass_controller.go:
```diff
@@ -0,0 +1,130 @@
+	log := r.Log.WithValues("runtimeclass", req.NamespacedName.Name)
+
+	runtimeClass := &nodeapi.RuntimeClass{}
+	err := r.Get(ctx, req.NamespacedName, runtimeClass)
```
Sorry for my ignorance...don't we need to filter by runtimeclass that we want to reconcile (kata-cc, kata-remote...etc) only

gkurz replied:
Yes we should. Good catch @wainersm !

beraldoleal replied:
Yep. I would just suggest may we don' t use the class names, to avoid having this "hardcoded" here. We could just set the SetupWithManager to filter by the finalizer name, if its our finalizer then reconcile.

I just pushed this approach. Let me know what you guys think.

wainersm replied:
Is https://github.com/openshift/sandboxed-containers-operator/pull/836/commits/9ff266d4d5829b49443c755a11bdecac722dd6a1#diff-2804d62dab597d3653f1d2640f5fa27727e60272d41de20713c4fea2fc981c19R134 that ensures it is filtered by the finalizer name?

gkurz replied:
Clever ! :smile:

wainersm commented:
@beraldoleal finished my review although I don't know much about the OSC operator (mainly the reconcile parts). So I don't know what is missing (if any) and/or what could be implemented better (if any).

wainersm reviewed (approved):
@beraldoleal , again, I'm not too versed on the operator but those changes looks good to me \o/

beraldoleal commented:
> @beraldoleal , again, I'm not too versed on the operator but those changes looks good to me \o/

"not too versed", but spotted a few good problems. thanks again.

beraldoleal commented:
@gkurz , @bpradipt... any other comments here? could we merge?

bpradipt reviewed (approved):
/lgtm
Thanks @beraldoleal

beraldoleal commented:
Discussed with @gkurz and got green light on slack. merged. thanks.
